File: solutions/0/0/1/6/p0016_s0.c
/* Solution to Euler Problem 16: Power Digit Sum. */
#include "runner.h"

/* No native big integers in C: hold base**power as a little-endian array of
 * decimal digits (least-significant at index 0) and build it by multiplying by
 * base `power` times (schoolbook long multiplication), then sum the digits.
 * O(power * digits) = O(power^2). */
const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int base  = parse_int(argv[1]);
    int power = parse_int(argv[2]);

    /* We'll store the big number in a decimal digit array.
     * 2^10000 has at most ceil(10000 * log10(2)) + 1 ~ 3011 decimal digits.
     * Allocate generously. */
    int max_digits = (int)(power * 0.30103 + 10) + 10;
    unsigned char *digits = calloc((size_t)max_digits, sizeof(unsigned char));
    if (!digits) {
        fprintf(stderr, "Out of memory\n");
        { snprintf(_answer, sizeof _answer, "%lld", (long long)(-1)); return _answer; }
    }

    /* Start with 1 */
    digits[0] = 1;
    int len = 1;

    /* Repeatedly multiply by base, power times */
    for (int i = 0; i < power; i++) {
        int carry = 0;
        for (int j = 0; j < len; j++) {
            int val = digits[j] * base + carry;
            digits[j] = (unsigned char)(val % 10);
            carry = val / 10;
        }
        while (carry > 0) {
            digits[len++] = (unsigned char)(carry % 10);
            carry /= 10;
        }
    }

    long long sum = 0;
    for (int i = 0; i < len; i++) {
        sum += digits[i];
    }

    free(digits);
    { snprintf(_answer, sizeof _answer, "%lld", (long long)(sum)); return _answer; }
}

File: solutions/0/0/1/6/p0016_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 16: Power Digit Sum [Level 0]. """
from __future__ import annotations

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Sum the decimal digits of base**power, computed with Python's built-in
    arbitrary-precision integers; an O(D) pass over its D ~ power*log10(base) digits."""
    base = runner.parse_int(args[0])
    power = runner.parse_int(args[1])

    return str(sum((int(i) for i in str(base**power))))


if __name__ == "__main__":
    raise SystemExit(solve())
